tests/mpi/periodicity_04.cc

```diff
@@ -0,0 +1,343 @@
+
+  FE_Q<dim> fe(1);
+  DoFHandler<dim> dof_handler(triangulation);
+  dof_handler.distribute_dofs(fe);
+
+  ConstraintMatrix      constraints;
+
+  IndexSet locally_owned_dofs = dof_handler.locally_owned_dofs();
+  IndexSet locally_relevant_dofs;
+  DoFTools::extract_locally_relevant_dofs (dof_handler, locally_relevant_dofs);
+  
+  constraints.reinit(locally_relevant_dofs);
+  {
+    std::vector<GridTools::PeriodicFacePair<typename DoFHandler<dim>::cell_iterator> >
+      periodicity_vector;
+
+    GridTools::collect_periodic_faces (dof_handler, 42, 43, dim-1, periodicity_vector);
+
+    DoFTools::make_periodicity_constraints<DoFHandler<dim> >
+      (periodicity_vector, constraints);
+  }
+  constraints.close ();
+
+  unsigned int myid = Utilities::MPI::this_mpi_process(MPI_COMM_WORLD);
+  std::ofstream file((std::string("dat.") + Utilities::int_to_string(myid)).c_str());
+  constraints.print(file);
+  file.close();
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  if (myid==0)
+    {
+      for (unsigned int i=0; i<Utilities::MPI::n_mpi_processes(MPI_COMM_WORLD); ++i)
+        cat_file((std::string("dat.") + Utilities::int_to_string(i)).c_str());
+    }
+
+  unsigned int n_local_constraints =0 ;
+
+  std::map<types::global_dof_index, Point<dim> > support_points;
+  DoFTools::map_dofs_to_support_points (MappingQ1<dim>(), dof_handler, support_points);
+  IndexSet constraints_lines = constraints.get_local_lines();
+
+  for (unsigned int i=0; i<constraints_lines.n_elements(); ++i)
+  {
+    const unsigned int line = constraints_lines.nth_index_in_set(i);
+    if (constraints.is_constrained(line))
+    {
+      const std::vector<std::pair<types::global_dof_index, double > >* entries 
+        = constraints.get_constraint_entries(line);
+      Assert(entries->size()==1, ExcInternalError());
+      const Point<dim> point1 = support_points[line];
+      const Point<dim> point2 = support_points[(*entries)[0].first];
+      Tensor<1, dim> difference = point1-point2;
+      difference[dim-1] = 0.;
+      AssertThrow(difference.norm()<1.e-9, ExcInternalError());
+      if (locally_owned_dofs.is_element(line))
+        ++n_local_constraints;
+    }
+  }
+  const unsigned int n_constraints 
+    = Utilities::MPI::sum(n_local_constraints, MPI_COMM_WORLD);
+  const unsigned int n_expected_constraints = Utilities::fixed_int_power<9,dim-1>::value;
+  if (myid==0)
+    deallog << "n_constraints: " << n_constraints 
+            << " n_expected_constraints: " << n_expected_constraints
+            << std::endl;
+  AssertThrow(n_constraints == n_expected_constraints, 
+              ExcDimensionMismatch(n_constraints, n_expected_constraints));
+}
+        
+int main (int argc, char *argv[])
+{
+  try
+    {
+      using namespace dealii;
+
+      Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
+      deallog.depth_console (0);
+
+      if (Utilities::MPI::this_mpi_process (MPI_COMM_WORLD)==0)
+        {
+          std::ofstream logfile("output");
```

Comment:
I often prefer ```  MPILogInitAll log;``` which will capture output of all processes and makes ```main()``` much simpler. See ```mpi/tria_01.cc``` for an example.